src/main/java/com/google/finapp/SpannerDaoJDBCImpl.java

```diff
@@ -0,0 +1,191 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.finapp;
+
+import com.google.cloud.ByteArray;
+import com.google.inject.Inject;
+import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Arrays;
+
+final class SpannerDaoJDBCImpl implements SpannerDaoInterface {
+
+  private final String projectId;
+  private final String databaseId;
+  private final String instanceId;
+  private final String connectionUrl;
+
+  @Inject
+  SpannerDaoJDBCImpl(
+      @ArgsModule.SpannerProjectId String spannerProjectId,
+      @ArgsModule.SpannerInstanceId String spannerInstanceId,
+      @ArgsModule.SpannerDatabaseId String spannerDatabaseId) {
+    this.projectId = spannerProjectId;
+    this.databaseId = spannerDatabaseId;
+    this.instanceId = spannerInstanceId;
+    String emulatorHost = System.getenv("SPANNER_EMULATOR_HOST");
+    if (emulatorHost != null) {
+      // connect to emulator
+      this.connectionUrl =
+          String.format(
+              "jdbc:cloudspanner://%s/projects/%s/instances/%s/databases/%s;usePlainText=true",
+              emulatorHost, projectId, instanceId, databaseId);
+    } else {
+      // connect to Cloud Spanner
+      this.connectionUrl =
+          String.format(
+              "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
+              projectId, instanceId, databaseId);
+    }
+  }
+
+  public void createCustomer(ByteArray customerId, String name, String address)
+      throws SpannerDaoException {
+    try {
+      try (Connection connection = DriverManager.getConnection(this.connectionUrl)) {
+        try (PreparedStatement ps =
```

Comment:
A few questions:
1. Why do you need the outer-most try?
2. Also have you considered defining multiple resources with a single try-with-resources, e.g: try this:
```
try (Connection connection = DriverManager.getConnection(this.connectionUrl);
       PreparedStatement ps = connection....blahblah...) {
```
same comment applies for other methods.